Bluetooth: L2CAP: simplify codepath for SDU TX

jori-nordic opened the pull request:
See commits.

TLDR: 
- always send SDUs from workqueue context
- ~remove~ fix support for net_buf_frags

jori-nordic committed:
Bluetooth: L2CAP: single codepath for sending SDUs

The purpose of this change is to help reason about what happens when a
application sends an SDU.

After this patch, the SDU is always added the channel's TX queue.
The queue is always processed from system workqueue context.

This change also enforces the application to reserve
`BT_L2CAP_SDU_CHAN_SEND_RESERVE` bytes in the buffer it passes to L2CAP.
This is to avoid unnecessary fragmentation and code complexity.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic commented:
Stacked on #67205 . Will get out of draft when merged.

jori-nordic committed:
tests: Bluetooth: use unique IDs for l2cap bsim tests

If $BASH_SOURCE is the empty string, then those two tests end up with
the same simulation ID. That's not a good time.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -579,22 +579,20 @@ int bt_l2cap_chan_disconnect(struct bt_l2cap_chan *chan);
  *
  *  When sending L2CAP data over an LE connection the application is sending
- *  L2CAP SDUs. The application can optionally reserve
+ *  L2CAP SDUs. The application should reserve
```
When providing suggestions (like `should`) I find it helpful to also explain _why_. 

Why should the application reserve this? What happens if it does/doesn't?

The migration guide now also states that it is required, but that does not fit with "should".

Thalley replied:
@jori-nordic Should this be changed to a `shall`?

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -923,47 +922,64 @@ static struct net_buf *l2cap_chan_le_get_tx_buf(struct bt_l2cap_le_chan *ch)
-		int sent = l2cap_tx_meta_data(buf)->sent;
+	while ((buf = l2cap_chan_le_get_tx_buf(le_chan))) {
+		/* Here buf is either:
+		 * - a half-sent SDU le_chan->tx_buf
```
```suggestion
		 * - a partially-sent SDU le_chan->tx_buf
```

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -923,47 +922,64 @@ static struct net_buf *l2cap_chan_le_get_tx_buf(struct bt_l2cap_le_chan *ch)
+				 * the current in-progress SDU. We will attempt
+				 * to resume sending it later.
+				 */
+				LOG_DBG("can't send right now (chan %p buf %p)", le_chan, buf);
```
```suggestion
				LOG_DBG("Cannot send right now (chan %p buf %p)", le_chan, buf);
```

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -923,47 +922,64 @@ static struct net_buf *l2cap_chan_le_get_tx_buf(struct bt_l2cap_le_chan *ch)
+						K_MSEC(CONFIG_BT_L2CAP_RESCHED_MS));
 			} else {
-				l2cap_tx_buf_destroy(ch->chan.conn, buf, sent);
+				LOG_WRN("couldn't send (err %d), dropping buf %p", err, buf);
```
```suggestion
				LOG_WRN("Failed to send (err %d), dropping buf %p", err, buf);
```

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1816,79 +1832,38 @@ static void le_disconn_rsp(struct bt_l2cap *l2cap, uint8_t ident,
-		net_buf_reserve(seg, BT_L2CAP_CHAN_SEND_RESERVE);
-		return seg;
+		if (!seg) {
+			LOG_WRN("app failed to allocate seg");
```
```suggestion
			LOG_WRN("App failed to allocate seg");
```

Thalley commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1816,79 +1832,38 @@ static void le_disconn_rsp(struct bt_l2cap *l2cap, uint8_t ident,
-	seg = l2cap_alloc_seg(buf, ch);
-
 	if (!seg) {
+		LOG_DBG("failed to allocate seg for buf %p", buf);
```
```suggestion
		LOG_DBG("Failed to allocate seg for buf %p", buf);
```

Minor general comment: Usually log statements start with a capital letter

Thalley commented on tests/bsim/bluetooth/host/l2cap/credits/src/main.c:
```diff
@@ -23,7 +23,8 @@ CREATE_FLAG(flag_l2cap_connected);
 
 /* Only one SDU transmitted or received at a time */
-NET_BUF_POOL_DEFINE(sdu_pool, 1, L2CAP_MTU, CONFIG_BT_CONN_TX_USER_DATA_SIZE, NULL);
+NET_BUF_POOL_DEFINE(sdu_pool, 1, L2CAP_MTU + BT_L2CAP_SDU_HDR_SIZE,
```
```suggestion
NET_BUF_POOL_DEFINE(sdu_pool, 1, BT_L2CAP_SDU_BUF_SIZE(L2CAP_MTU),
```

Thalley replied:
Does it? 

`CONFIG_BT_L2CAP_TX_MTU` is usually used with `BT_L2CAP_SDU_BUF_SIZE`, so that is without the header, right? 

`L2CAP_MTU` is a multiple of `SDU_LEN` which is a multiple of `L2CAP_MPS` which is defined as `CONFIG_BT_L2CAP_TX_MTU`

Given that they are multiples, and not using a constant addition to the size, I would assume that none of the above are including the header size (granted, due the multiples, they can certainly fit it, but none of them are adding `BT_L2CAP_HDR_SIZE`). 

If the intention is that `L2CAP_MTU` contains the size of the header, then I feel the above macro definitions are not clear about that (at least not to me). That being said, it is also unclear to me why we are multiplying by the MTU by 4 (2 * 2) in the first place, so I might just be missing something here.

jori-nordic replied:
Yes, you're right my bad. MTU & MPS both don't include headers. Updated.

Thalley reviewed (changes requested):
A few minor comments, otherwise LGTM

jori-nordic commented:
@rlubos FYI.
Do you know if IPSP / Bluetooth L2 net uses buffer fragments?

rlubos commented:
> Do you know if IPSP / Bluetooth L2 net uses buffer fragments?

I might be a bit out of context here, what do you mean by "buffer fragments"?

jori-nordic commented:
> > Do you know if IPSP / Bluetooth L2 net uses buffer fragments?
> 
> I might be a bit out of context here, what do you mean by "buffer fragments"?

this: https://github.com/jori-nordic/zephyr/blob/303b2492d69ed375e069eba09f4ad7f8588c382b/include/zephyr/net/buf.h#L2336

jori-nordic commented:
@rlubos is there a way to test the IP L2 transport without involving a linux machine?
Ie I want to exercise those network fragments, but only involve zephyr devices.

rlubos commented:
> @rlubos is there a way to test the IP L2 transport without involving a linux machine?

If you meant IPSP L2, the only valid testing scenario I know involes Linux host (as described in IPSP sample docs). But I don't claim to be BLE expert, i. e. I don't know if it's possible to test device vs device.

jori-nordic committed:
Bluetooth: L2CAP: fix net buf frags handling

Fix the handling of buffers with fragments.
Add a test to regression too.

The main user of this is IPSP.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

jori-nordic commented:
Since @rlubos told me we had at least one user of frags, I decided to support them properly with a test.

jori-nordic committed:
Bluetooth: L2CAP: stop stealing buffers from SDU pool

It seems like a nice idea at first, but leads to hard-to-debug
situations for the application.

The previous behavior can be implemented by the app by defining
`alloc_seg` and allocating from the same pool as `buf`.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

Thalley commented on tests/bsim/bluetooth/host/l2cap/frags/src/main.c:
```diff
@@ -0,0 +1,477 @@
+CREATE_FLAG(is_connected);
+CREATE_FLAG(flag_l2cap_connected);
+
+#define SMOL_SIZE 10
```
Should this have been

```suggestion
#define SMALL_SIZE 10
```

jori-nordic replied:
I'll un-meme it for you :)

Thalley replied:
Hah, I was wondering if that was the case (especially with `chonk`) - And while that is indeed fun, I think a less meme-oriented naming convention is better here :D

Thalley commented on tests/bsim/bluetooth/host/l2cap/frags/src/main.c:
```diff
@@ -0,0 +1,477 @@
+CREATE_FLAG(flag_l2cap_connected);
+
+#define SMOL_SIZE 10
+#define CHONK_SIZE 50
```
What is a `chonk` in this context?

jori-nordic replied:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/68428270/b9729ef9-ca44-4d01-94ab-3a5aa3837bf0)

Thalley commented on tests/bsim/bluetooth/host/l2cap/frags/src/main.c:
```diff
@@ -0,0 +1,477 @@
+
+struct test_ctx {
+	struct bt_l2cap_le_chan le_chan;
+	size_t tx_left;
```
I got confused by this naming, as I was thinking of left / right. 

Consider using
```suggestion
	size_t tx_remaining;
```

Thalley commented on tests/bsim/bluetooth/host/l2cap/frags/src/main.c:
```diff
@@ -0,0 +1,477 @@
+	WAIT_FOR_FLAG_UNSET(is_connected);
+}
+
+#define BT_LE_ADV_CONN_NAME_OT BT_LE_ADV_PARAM(BT_LE_ADV_OPT_CONNECTABLE | \
+					    BT_LE_ADV_OPT_USE_NAME |	\
+					    BT_LE_ADV_OPT_ONE_TIME,	\
+					    BT_GAP_ADV_FAST_INT_MIN_2, \
+					    BT_GAP_ADV_FAST_INT_MAX_2, NULL)
```
Some odd indentation here

jori-nordic replied:
will fix

jori-nordic commented:
> @jori-nordic Should this be changed to a shall?

huh I thought I had made the change. another one lost in the sands of ~time~ git rebase.

jori-nordic commented:
@Thalley there were some `net_buf_ref` issues I had to debug this morning, should be good for review (if CI passes) now.

Thalley commented:
> > @jori-nordic Should this be changed to a shall?
> 
> huh I thought I had made the change. another one lost in the sands of ~time~ git rebase.

Yeah, I feel like I saw it fixed earlier as well, and then somehow reverted :s

alwa-nordic commented on doc/releases/migration-guide-3.6.rst:
```diff
@@ -212,6 +212,10 @@ Bluetooth
   Any pointer to a UUID must be prefixed with `const`, otherwise there will be a compilation warning.
   For example change ``struct bt_uuid *uuid = BT_UUID_DECLARE_16(xx)`` to
   ``const struct bt_uuid *uuid = BT_UUID_DECLARE_16(xx)``. (:github:`66136`)
+* The :c:func:`bt_l2cap_chan_send` API now doesn't return the number of bytes sent.
```
This should go in the change log. Also add that it never returned anything useful.

alwa-nordic commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -579,22 +579,28 @@ int bt_l2cap_chan_disconnect(struct bt_l2cap_chan *chan);
- *  to allocate buffers from the original buffer pool of the L2CAP SDU before
- *  using the stacks own buffer pool.
+ *
+ *  The application can use the BT_L2CAP_SDU_BUF_SIZE() helper to correctly size
+ *  the buffers for the outgoing buffer pool.
```
```suggestion
 *  The application can use the BT_L2CAP_SDU_BUF_SIZE() helper to correctly size
 *  the buffers to account for the reserved headroom.
```

alwa-nordic commented on include/zephyr/bluetooth/l2cap.h:
```diff
@@ -579,22 +579,28 @@ int bt_l2cap_chan_disconnect(struct bt_l2cap_chan *chan);
  *
- *  @return Bytes sent in case of success or negative value in case of error.
+ *  @return 0 in case of success or negative value in case of error. See below.
+ *  @return -EINVAL if `buf` or `chan` is NULL.
```
`retval`

alwa-nordic commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -923,47 +922,64 @@ static struct net_buf *l2cap_chan_le_get_tx_buf(struct bt_l2cap_le_chan *ch)
 	/* Resume tx in case there are buffers in the queue */
-	while ((buf = l2cap_chan_le_get_tx_buf(ch))) {
-		int sent = l2cap_tx_meta_data(buf)->sent;
+	while ((buf = l2cap_chan_le_get_tx_buf(le_chan))) {
```
Separate commit

alwa-nordic commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1816,79 +1832,38 @@ static void le_disconn_rsp(struct bt_l2cap *l2cap, uint8_t ident,
 
-	/* Use the dedicated segment callback if registered */
 	if (ch->chan.ops->alloc_seg) {
+		/* Use the dedicated segment callback if registered */
```
"Use the above user-defined al-locator."

alwa-nordic commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1816,79 +1832,38 @@ static void le_disconn_rsp(struct bt_l2cap *l2cap, uint8_t ident,
-		goto segment;
+	if (!seg) {
+		/* Try to use original pool if possible */
+		seg = net_buf_alloc(pool, K_NO_WAIT);
```
/* Y O I N K */

alwa-nordic commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -1980,41 +1955,60 @@ static bool test_and_dec(atomic_t *target)
+			return -EAGAIN;
+		}
+
+		/* Don't send more that TX MPS */
```
/ðan/

alwa-nordic commented on subsys/bluetooth/host/l2cap.c:
```diff
@@ -2057,61 +2057,56 @@ static int l2cap_chan_le_send(struct bt_l2cap_le_chan *ch,
-		return -EMSGSIZE;
-	}
-
+	/* Those are _not_ HCI fragments nor L2CAP segments. They are the
```
"The `frag` here does not refer to"

alwa-nordic reviewed (changes requested):
Notes from offline review

jori-nordic commented:
I will split some stuff out, open multiple PRs instead tomorrow.
